File: algorithms/2044-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.go
package problem2044


func countMaxOrSubsets(nums []int) int {
	maxOr, cnt := 0, 0

	var dfs func(pos int, orVal int)
	dfs = func(pos int, orVal int) {
		if pos == len(nums) {
			if orVal > maxOr {
				maxOr = orVal
				cnt = 1
			} else if orVal == maxOr {
				cnt++
			}
			return
		}

		dfs(pos + 1, orVal)
		dfs(pos + 1, orVal | nums[pos])
	}

	dfs(0, 0)
	return cnt
}

File: algorithms/2044-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.py
"""
2044. count-number-of-maximum-bitwise-or-subsets

统计按位或能得到最大值的子集数目
"""

from typing import List


def count_max_or_subsets(nums: List[int]) -> int:
    """
    回溯
    """
    max_or, cnt = 0, 0

    def dfs(pos: int, or_val: int) -> None:
        if pos == len(nums):
            nonlocal max_or, cnt
            if or_val > max_or:
                max_or, cnt = or_val, 1
            elif or_val == max_or:
                cnt += 1
            return
        dfs(pos + 1, or_val | nums[pos])
        dfs(pos + 1, or_val)

    dfs(0, 0)
    return cnt
